reference/src/main/scala/akka/stream/alpakka/reference/settings.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.reference
+
+// rename Java imports if the name clashes with the Scala name
+import java.time.{Duration => JavaDuration}
+import java.util.function.Predicate
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.duration._
+
+/**
+ * Settings class constructor is private and not exposed as API.
+ * Adding or removing arguments to methods with default values is not binary
+ * compatible. However, since the constructor is private, it will be possible
+ * to add or remove attributes without introducing binary incompatibilities.
+ */
+final class SourceSettings private (
+    val clientId: Option[String] = None,
+    val authentication: Authentication = Authentication.None,
+    val pollInterval: Duration = 5.seconds
+) {
+
+  /**
+   * Immutable setter which can be used from both Java and Scala, even if the
+   * attribute is stored in a Scala specific class.
+   */
+  def withClientId(clientId: String): SourceSettings = copy(clientId = Some(clientId))
+
+  /**
+   * Separate setters for every attribute enables easy evolution of settings classes by allowing
+   * deprecation and addition of attributes.
+   */
+  def withAuthentication(authentication: Authentication): SourceSettings = copy(authentication = authentication)
+
+  /**
+   * For attributes that uses Java or Scala specific classes, a setter is added for both APIs.
+   */
+  def withPollInterval(pollInterval: Duration): SourceSettings = copy(pollInterval = pollInterval)
+
+  /**
+   * Java API
+   *
+   * Start method documentation text with "Java API" to make it easy to notice
+   * Java specific methods when browsing generated API documentation.
+   */
+  def withPollInterval(pollInterval: JavaDuration): SourceSettings = {
+    import scala.compat.java8.DurationConverters._
+    copy(pollInterval = pollInterval.toScala)
+  }
+
+  /**
+   * Private copy method for internal use only.
+   */
+  private def copy(clientId: Option[String] = clientId,
+                   authentication: Authentication = authentication,
+                   pollInterval: Duration = pollInterval) =
+    new SourceSettings(clientId, authentication, pollInterval)
+
+  override def toString: String =
+    s"""SourceSettings(
+       |  clientId       = $clientId
+       |  authentication = $authentication
+       |  pollInterval   = $pollInterval
+       |)""".stripMargin
+}
+
+object SourceSettings {
+
+  /**
+   * Factory method for Scala.
+   */
+  def apply(): SourceSettings = new SourceSettings()
```

Comment:
If there are settings without good defaults (eg. connection details), these should appear in `apply/create`.